// @alexa-skill-components/feedback-elicitation/dist/util.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __values = (this && this.__values) || function(o) {
    var s = typeof Symbol === "function" && Symbol.iterator, m = s && o[s], i = 0;
    if (m) return m.call(o);
    if (o && typeof o.length === "number") return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
    throw new TypeError(s ? "Object is not iterable." : "Symbol.iterator is not defined.");
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.getResolvedApiSlotIds = exports.getSlotResolvedId = exports.getResolvedApiSlotValues = exports.getSlotResolvedValue = exports.getApiSlots = exports.getApiArguments = exports.isApiRequestPrefix = exports.isApiRequest = exports.getAPIName = void 0;
var ask_sdk_core_1 = require("ask-sdk-core");
var _ = __importStar(require("lodash"));
var getAPIName = function (handlerInput) {
    var _a;
    return (_a = handlerInput.requestEnvelope.request.apiRequest) === null || _a === void 0 ? void 0 : _a.name;
};
exports.getAPIName = getAPIName;
/**
 * Helper method to find if a request is a API request, optionally for a specific API
 */
var isApiRequest = function (handlerInput, apiName) {
    var isApiRequest = (0, ask_sdk_core_1.getRequestType)(handlerInput.requestEnvelope) === 'Dialog.API.Invoked';
    if (apiName) {
        return isApiRequest && (0, exports.getAPIName)(handlerInput) === apiName;
    }
    else {
        return isApiRequest;
    }
};
exports.isApiRequest = isApiRequest;
var isApiRequestPrefix = function (handlerInput, apiName) {
    var _a;
    var isApiRequest = (0, ask_sdk_core_1.getRequestType)(handlerInput.requestEnvelope) === 'Dialog.API.Invoked';
    return isApiRequest && (((_a = (0, exports.getAPIName)(handlerInput)) === null || _a === void 0 ? void 0 : _a.startsWith(apiName)) || false);
};
exports.isApiRequestPrefix = isApiRequestPrefix;
/**
 * Helper method to get API request entity from the request envelope.
 */
var getApiArguments = function (handlerInput) {
    var _a;
    return ((_a = handlerInput.requestEnvelope.request.apiRequest) === null || _a === void 0 ? void 0 : _a.arguments) || {};
};
exports.getApiArguments = getApiArguments;
/**
 * Helper method to get API resolved entity from the request envelope.
 */
var getApiSlots = function (handlerInput) {
    var _a;
    return ((_a = handlerInput.requestEnvelope.request.apiRequest) === null || _a === void 0 ? void 0 : _a.slots) || {};
};
exports.getApiSlots = getApiSlots;
var getSlotResolvedValue = function (slot) {
    var firstAuthority = _.first(_.get(slot, 'resolutions.resolutionsPerAuthority'));
    var firstAuthorityValue = _.first(_.get(firstAuthority, 'values'));
    return _.get(firstAuthorityValue, 'value.name');
};
exports.getSlotResolvedValue = getSlotResolvedValue;
var getResolvedApiSlotValues = function (handlerInput) {
    var e_1, _a;
    var slots = (0, exports.getApiSlots)(handlerInput);
    var resolvedSlotValues = {};
    try {
        for (var _b = __values(Object.keys(slots)), _c = _b.next(); !_c.done; _c = _b.next()) {
            var slotName = _c.value;
            resolvedSlotValues[slotName] = (0, exports.getSlotResolvedValue)(slots[slotName]);
        }
    }
    catch (e_1_1) { e_1 = { error: e_1_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_1) throw e_1.error; }
    }
    return resolvedSlotValues;
};
exports.getResolvedApiSlotValues = getResolvedApiSlotValues;
var getSlotResolvedId = function (slot) {
    var firstAuthority = _.first(_.get(slot, 'resolutions.resolutionsPerAuthority'));
    var firstAuthorityValue = _.first(_.get(firstAuthority, 'values'));
    return _.get(firstAuthorityValue, 'value.id');
};
exports.getSlotResolvedId = getSlotResolvedId;
var getResolvedApiSlotIds = function (handlerInput) {
    var e_2, _a;
    var slots = (0, exports.getApiSlots)(handlerInput);
    var resolvedSlotIds = {};
    try {
        for (var _b = __values(Object.keys(slots)), _c = _b.next(); !_c.done; _c = _b.next()) {
            var slotName = _c.value;
            resolvedSlotIds[slotName] = (0, exports.getSlotResolvedId)(slots[slotName]);
        }
    }
    catch (e_2_1) { e_2 = { error: e_2_1 }; }
    finally {
        try {
            if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
        }
        finally { if (e_2) throw e_2.error; }
    }
    return resolvedSlotIds;
};
exports.getResolvedApiSlotIds = getResolvedApiSlotIds;
