// @alexa-skill-components/feedback-elicitation/dist/handlers/save-rating-handler.js
"use strict";
// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
// SPDX-License-Identifier: LicenseRef-.amazon.com.-AmznSL-1.0
// Licensed under the Amazon Software License  http://aws.amazon.com/asl/
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SaveRatingRequestHandler = void 0;
var util = __importStar(require("../util"));
var config_1 = require("../config");
var log_rating_recorder_1 = require("../recorders/log-rating-recorder");
/* *
 * SaveRatingRequestHandler records the rating and returns the response to the skill.
 * It can be used to handle requests to the action specified by the saveRatingAction argument to the elicitRating reusable dialog.
 * This handler will be triggered when feedback is collected: rating
 * Response contains the rating details which maps to rating type in ACDL and display for APL template
 * */
var SaveRatingRequestHandler = /** @class */ (function () {
    function SaveRatingRequestHandler(apiName, ratingHandler, shouldEndSession) {
        if (apiName === void 0) { apiName = "".concat(config_1.feedbacknamespace, ".defaultSaveFeedbackAction"); }
        if (ratingHandler === void 0) { ratingHandler = new log_rating_recorder_1.LogRatingRecorder(); }
        if (shouldEndSession === void 0) { shouldEndSession = true; }
        this.apiName = apiName;
        if ("handleRating" in ratingHandler) {
            this.ratingHandler = ratingHandler;
        }
        else {
            this.ratingHandler = {
                "handleRating": ratingHandler
            };
        }
        this.shouldEndSession = shouldEndSession;
    }
    SaveRatingRequestHandler.prototype.canHandle = function (handlerInput) {
        return util.isApiRequest(handlerInput, this.apiName);
    };
    SaveRatingRequestHandler.prototype.handle = function (handlerInput) {
        var rating = util.getApiArguments(handlerInput).rating;
        this.ratingHandler.handleRating(rating);
        //response maps to rating type in ACDL
        var apiResponse = {
            rating: rating
        };
        /**
         * This handler acts as the entry point for your skill, routing all request and response
         * payloads to the handlers above. Make sure any new handlers or interceptors you've
         * defined are included below. The order matters - they're processed top to bottom
         * */
        return handlerInput.responseBuilder
            .withApiResponse(apiResponse)
            .withShouldEndSession(this.shouldEndSession)
            .getResponse();
    };
    return SaveRatingRequestHandler;
}());
exports.SaveRatingRequestHandler = SaveRatingRequestHandler;
